app/src/main/java/com/mathewsachin/fategrandautomata/ui/support_img_namer/SupportImageNamer.kt

```diff
@@ -107,17 +111,22 @@ fun showSupportImageNamer(UI: ScriptRunnerUserInterface, storageDirs: StorageDir
 
     val entryList = getSupportEntries(frame, storageDirs)
 
-    showOverlayDialog(context) {
-        setCancelable(false)
-            .setTitle(UI.Service.getString(R.string.support_img_namer_title))
-            .setView(content)
-            .setPositiveButton(UI.Service.getString(R.string.support_img_namer_done)) { dialog, _ ->
-                if (entryList.all { it.isValid() }) {
-                    if (entryList.all { it.rename() }) {
-                        dialog.dismiss()
+    suspendCancellableCoroutine<Unit> { coroutine ->
```

Comment:
I can't wrap my head around this. What does suspendCancellableCoroutine do in this case? I tried reading about it, but the only thing I can find is that suspendCancellableCoroutine is supposed to replace callbacks.

Reply:
It turns callbacks into coroutines. The benefit is instead of passing a completion callback to `supportImageNamer` function. I can call it as if I'm calling a sequential function from the main thread, but at the same time not blocking the main thread. And, once the dialog is closed by the user, the rest of the code runs.

Reply:
If I understand correctly, the `showOverlayDialog` is executed immediately when this line is reached and then when either the positive or the negative button is hit, `coroutine.resume()` is called.

`coroutine` in this case is the method `showSupportImageNamer`, which has nothing left to do, so the the method which called `showSupportImageNamer` is resumed.

Is that right?

Reply:
yes, that's how it works.
Kotlin coroutines are a bit different from async/await in C# and JS in that the default behaviour is to execute code sequentially, unlike C#/JS where you have to `await` for sequential execution.